tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -869,6 +1055,90 @@ def _search_instantiations(
 
             self._search_instantiations(fn, via=sig)
 
+    def find_decl_by_unique_id(self, unique_id) -> Optional[NodeAndRange]:
+        """
+        Traverse the (location-key'ed) definition segment table and
+        find the definition segment with the given ID, or none.
+        """
+        return next(
+            (
+                decl
+                for decl in self.def_segments.elts
+                if decl.node.unique_id() == unique_id
+            ),
+            None,
+        )
+
+    def find_instantiation_by_unique_id(
+        self, unique_id
+    ) -> Optional[Tuple[NodeAndRange, chapel.TypedSignature]]:
+        """
+        Traverse the (location-key'ed) definition segment table and
+        find the definition segment with the given ID, or none.
+        """
+        return next(
+            (
+                decl
+                for decl in self.instantiation_segments.elts
+                if decl[0].node.unique_id() == unique_id
+            ),
+            None,
+        )
+
+    def update_call_segments_from_instantiations(self, rng: Range):
+        self.call_segments.clear_range(rng)
+        start, end = self.instantiation_segments._get_segment_range(rng)
+
+        segments_for_elt = {}
+
+        def process_instantiation_segment(value_at_segment, elt_idx):
+            if elt_idx in segments_for_elt:
+                return segments_for_elt[elt_idx]
+
+            inst, sig = value_at_segment
+            calls_in_inst = PositionList[ResolvedPair](lambda x: x.ident.rng)
+
+            for node in chapel.preorder(inst.node):
+                if not isinstance(node, chapel.FnCall):
+                    continue
+
```

Comment:
why not  `chapel.each_matching(inst.node, chapel.FnCall, iterator=chapel.postorder`?